lib/elixir/lib/json.ex

```diff
@@ -0,0 +1,281 @@
+          [] ->
+            fields -- [:__struct__ | except]
+
+          error_keys ->
+            raise ArgumentError,
+                  "unknown struct fields #{inspect(error_keys)} specified in :except. Expected one of: " <>
+                    "#{inspect(fields -- [:__struct__])}"
+        end
+
+      true ->
+        fields -- [:__struct__]
+    end
+  end
+
+  @doc """
+  A function invoked to encode the given term.
+  """
+  def encode(term, encoder)
+end
+
+defimpl JSON.Encoder, for: Atom do
+  def encode(value, encoder) do
+    case value do
+      nil -> "null"
+      true -> "true"
+      false -> "false"
+      _ -> encoder.(Atom.to_string(value), encoder)
+    end
+  end
+end
+
+defimpl JSON.Encoder, for: BitString do
+  def encode(value, _encoder) do
+    :elixir_json.encode_binary(value)
+  end
+end
+
+defimpl JSON.Encoder, for: List do
+  def encode(value, encoder) do
+    :elixir_json.encode_list(value, encoder)
+  end
+end
+
+defimpl JSON.Encoder, for: Integer do
+  def encode(value, _encoder) do
+    :elixir_json.encode_integer(value)
+  end
+end
+
+defimpl JSON.Encoder, for: Float do
+  def encode(value, _encoder) do
+    :elixir_json.encode_float(value)
+  end
+end
+
+defimpl JSON.Encoder, for: Map do
+  def encode(value, encoder) do
+    :elixir_json.encode_map(value, encoder)
+  end
+end
+
+defmodule JSON do
+  @moduledoc ~S"""
+  JSON encoding and decoding.
+
+  Both encoder and decoder fully conform to [RFC 8259](https://tools.ietf.org/html/rfc8259) and
+  [ECMA 404](https://ecma-international.org/publications-and-standards/standards/ecma-404/)
+  standards.
+
+  ## Encoding
+
+  Elixir built-in data structures are encoded to JSON as follows:
+
+  | **Elixir**             | **JSON** |
+  |------------------------|----------|
+  | `integer() \| float()` | Number   |
+  | `true \| false `       | Boolean  |
+  | `nil`                  | Null     |
+  | `binary()`             | String   |
+  | `atom()`               | String   |
+  | `list()`               | Array    |
+  | `%{binary() => _}`     | Object   |
+  | `%{atom() => _}`       | Object   |
+  | `%{integer() => _}`    | Object   |
+
+  You may also implement the `JSON.Encoder` protocol for custom data structures.
+
+  ## Decoding
+
+  Elixir built-in data structures are decoded from JSON as follows:
+
+  | **JSON** | **Elixir**             |
+  |----------|------------------------|
+  | Number   | `integer() \| float()` |
+  | Boolean  | `true \| false`        |
+  | Null     | `nil`                  |
+  | String   | `binary()`             |
+  | Object   | `%{binary() => _}`     |
+
+  """
+
+  @moduledoc since: "1.18.0"
+
+  @doc ~S"""
+  Encodes the given term to JSON as a binary.
+
+  The second argument is a function that is recursively
+  invoked to encode a term.
+
+  ## Examples
+
+      iex> JSON.encode([123, "string", %{key: "value"}])
+      "[123,\"string\",{\"key\":\"value\"}]"
+
+  """
+  def encode(term, encoder \\ &encode_value/2) do
```

Comment:
What is the reason we went with a different API than [`Jason`](https://hexdocs.pm/jason/Jason.html#encode/2)?
What would a use case for another `encoder` be, perhaps we could provide an example?
The `:pretty` option from `Jason` is quite useful, it might be nice to have a way to add such options in the future.

Reply:
`encode` returning an ok tuple is not really reliable or consistent. Even in Jason, it has several scenarios where it may raise. So we went with an API that is compatible with prior art in Elixir, and that is `Base`

Also, if we want to be Erlang compatible, then the recursive encoder is the way to go. If you want pretty encoder in the future, you would do: `JSON.encode(term, &pretty_encoder/2)`.

Reply:
Makes perfect sense, thanks!